wasm/src/main/scala/ir2wasm/HelperFunctions.scala

```diff
@@ -466,6 +494,373 @@ object HelperFunctions {
+      "fromAncestorsLocal",
+      WasmRefType(WasmHeapType.Type(WasmArrayTypeName.typeDataArray))
+    )
+    val lenLocal = fctx.addLocal("len", WasmInt32)
+    val iLocal = fctx.addLocal("i", WasmInt32)
+
+    // if (fromTypeData eq typeData)
+    instrs += LOCAL_GET(fromTypeDataParam)
+    instrs += LOCAL_GET(typeDataParam)
+    instrs += REF_EQ
+    fctx.ifThen() {
+      // then return true
+      instrs += I32_CONST(I32(1))
+      instrs += RETURN
+    }
+
+    // "Tail call" loop for diving into array component types
+    fctx.loop(WasmInt32) { loopForArrayLabel =>
+      // switch (typeData.kind)
+      fctx.switch() { () =>
+        // typeData.kind
+        instrs += LOCAL_GET(typeDataParam)
+        instrs += STRUCT_GET(TypeIdx(WasmStructTypeName.typeData), kindIdx)
+      }(
+        // case anyPrimitiveKind => return false
+        (KindVoid to KindLastPrimitive).toList -> { () =>
+          instrs += I32_CONST(I32(0))
+          instrs += RETURN
+        },
+        // case KindArray => check that from is an array, recurse into component types
+        List(KindArray) -> { () =>
+          fctx.block() { fromComponentTypeIsNullLabel =>
+            // fromTypeData := fromTypeData.componentType; jump out if null
+            instrs += LOCAL_GET(fromTypeDataParam)
+            instrs += STRUCT_GET(TypeIdx(WasmStructTypeName.typeData), componentTypeIdx)
+            instrs += BR_ON_NULL(fromComponentTypeIsNullLabel)
+            instrs += LOCAL_SET(fromTypeDataParam)
+
+            // typeData := ref.as_non_null typeData.componentType (OK because KindArray)
+            instrs += LOCAL_GET(typeDataParam)
+            instrs += STRUCT_GET(TypeIdx(WasmStructTypeName.typeData), componentTypeIdx)
+            instrs += REF_AS_NOT_NULL
+            instrs += LOCAL_SET(typeDataParam)
+
+            // loop back ("tail call")
+            instrs += BR(loopForArrayLabel)
+          }
+
+          // return false
+          instrs += I32_CONST(I32(0))
+          instrs += RETURN
+        },
+        // case KindObject => return (fromTypeData.kind > KindLastPrimitive)
+        List(KindObject) -> { () =>
+          instrs += LOCAL_GET(fromTypeDataParam)
+          instrs += STRUCT_GET(TypeIdx(WasmStructTypeName.typeData), kindIdx)
+          instrs += I32_CONST(I32(KindLastPrimitive))
+          instrs += I32_GT_U
+          instrs += RETURN
+        },
+        // case KindCloneable | KindCloneableSerializable =>
+        //   if from is an array, return true, otherwise fall through
+        List(KindCloneable, KindSerializable) -> { () =>
+          // if (fromTypeData.componentType ne null)
+          instrs += LOCAL_GET(fromTypeDataParam)
+          instrs += STRUCT_GET(TypeIdx(WasmStructTypeName.typeData), componentTypeIdx)
+          instrs += REF_IS_NULL
+          instrs += I32_EQZ
+          fctx.ifThen() {
+            // then return true
+            instrs += I32_CONST(I32(1))
+            instrs += RETURN
+          }
```

Comment:
Yes, it's because of that.

Your question prompted me to reconsider this section, and I actually found a better way to deal with those cases. Now I put `Cloneable` and `Serializable` in the `strictAncestors` of array type datas, so we don't need this special case at all anymore.

I also expanded a little bit in comments.

Is it clearer now?